### PR TITLE
Refs #373 -- Added TupleIn subqueries.

### PR DESCRIPTION
**Pull request by csirmazbendeguz**
#### Trac ticket number
ticket-373

#### Branch description
Added the `supports_tuple_in_subquery` database feature.

This enables developers to use `TupleIn` with subqueries, e.g. `WHERE (foo, bar) IN (SELECT foo, bar FROM baz)`.

---

This will be used by the feature requested by @LilyFoote (https://github.com/django/django/pull/18056#discussion_r1700379135) later. The feature is the ability to UPDATE objects with composite primary keys + a subquery filter in one go (without pre-selecting the `pk`s first).

e.g.:
```
UPDATE table
SET column = 'value'
WHERE (id1, id2) IN (SELECT id1, id2 FROM table)
```

See `SQLUpdateCompiler.pre_sql_setup` for more details (https://github.com/django/django/pull/18056/files#diff-f58de2deaccecd2d53199c5ca29e3e1050ec2adb80fb057cdfc0b4e6accdf14fR2128).

While all supported database backends support this operation, we need a feature flag so 3rd party database backends that don't support it can disable it (e.g. SQL Server) and fall back to pre-selecting IDs before update.

#### Checklist
- [X] This PR targets the `main` branch. 
- [X] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [X] I have checked the "Has patch" ticket flag in the Trac system.
- [X] I have added or updated relevant tests.
- [X] I have added or updated relevant docs, including release notes if applicable.
- [X] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/db/models/fields/tuple_lookups.py [hunk 1]
```diff
@@ -233,6 +240,23 @@ def check_rhs_elements_length_equals_lhs_length(self):
                 f"must have {len_lhs} elements each"
             )
 
+    def check_rhs_is_query(self):
+        if not (isinstance(self.rhs, Query)):
+            lhs_str = self.get_lhs_str()
+            raise ValueError(
+                f"{self.lookup_name!r} subquery lookup of {lhs_str} "
+                "must be a Query object"
+            )
```
Comment (LilyFoote) [inline django/db/models/fields/tuple_lookups.py:249]: We should probably include something specific about `self.rhs` in this error message. Currently we only have `self.lhs`.

File: django/db/models/fields/tuple_lookups.py [hunk 2]
```diff
@@ -233,6 +240,23 @@ def check_rhs_elements_length_equals_lhs_length(self):
...
+
+    def check_rhs_select_length_equals_lhs_length(self):
+        len_lhs = len(self.lhs)
+        if len(self.rhs.select) != len_lhs:
+            lhs_str = self.get_lhs_str()
+            raise ValueError(
+                f"{self.lookup_name!r} subquery lookup of {lhs_str} "
+                f"must have {len_lhs} fields"
+            )
```
Comment (LilyFoote) [inline django/db/models/fields/tuple_lookups.py:258]: Similarly here. Perhaps we can report `len(self.rhs.select)`?

File: django/db/models/fields/tuple_lookups.py (hunk 1 above)
Comment (csirmazbendeguz) [reply django/db/models/fields/tuple_lookups.py]: Alright, I added the class name to the error message 👍

File: django/db/models/fields/tuple_lookups.py (hunk 2 above)
Comment (csirmazbendeguz) [reply django/db/models/fields/tuple_lookups.py]: Okay, I added the `rhs` length to the error message 👍

File: django/db/models/fields/tuple_lookups.py [hunk 3]
```diff
@@ -271,6 +304,13 @@ def as_sqlite(self, compiler, connection):
 
         return root.as_sql(compiler, connection)
 
+    def as_subquery(self, compiler, connection):
+        if not connection.features.supports_tuple_in_subquery:
+            raise NotSupportedError(
+                "TupleIn subqueries are not supported on this database backend."
```
Comment (LilyFoote) [inline django/db/models/fields/tuple_lookups.py:310]: Can we mention which database backend that is? It's possible to configure Django to use more than one, in which case maybe the programmer meant to use a different one.

**Review by LilyFoote — APPROVED** [review]

**Comment by sarahboyce** [comment]
buildbot, test on oracle.

**Comment by sarahboyce** [comment]
> While all supported database backends support this operation, we need a feature flag so 3rd party database backends that don't support it can disable it (e.g. SQL Server) and fall back to pre-selecting IDs before update.

I've not seen us do this before. Doesn't mean it isn't something we do
If they set `supports_tuple_in_subquery = False` they get a `NotSupportedError` right? I'm not sure where the fallback behavior is

**Comment by csirmazbendeguz** [comment]
@sarahboyce , the idea is to
1. add subquery support to the TupleIn now
2. add the UPDATE subquery support in the [composite pk patch](https://github.com/django/django/pull/18056/files#diff-f58de2deaccecd2d53199c5ca29e3e1050ec2adb80fb057cdfc0b4e6accdf14fR2128)

so yes the feature flag is not utilized for fallback in this patch but it will be in the next patch

**Comment by csirmazbendeguz** [comment]
but if you think this is not the best approach then let me know what solution you would prefer - that's is why it's good to have these smaller PRs so we can discuss these smaller things in detail 😊

**Comment by sarahboyce** [comment]
> maybe we don't need the feature flag at all?

I don't know - maybe @charettes can confirm?

**Comment by charettes** [comment]
If no core backends requires it today I'd suggest starting without a feature switch for now. We've seen third-party app maintainers be pretty reactive about requesting ones when needed.

Thoughts on the above @timgraham?

**Comment by csirmazbendeguz** [comment]
Thanks @charettes - alright, let's not worry about this right now, I suppose it's always easier to add it later than to remove it. @sarahboyce , I removed the feature switch.

**Comment by timgraham** [comment]
The backends I maintain (CockroachDB and Snowflake) don't require the feature flag.

**Comment by LilyFoote** [comment]
> While all supported database backends support this operation, we need a feature flag so 3rd party database backends that don't support it can disable it (e.g. SQL Server)

Given the concern about SQL Server specifically, maybe we should check with the maintainers of [mssql-django](https://github.com/microsoft/mssql-django)? @mshan0 @absci do you have any thoughts here?

**Comment by absci** [comment]
> > While all supported database backends support this operation, we need a feature flag so 3rd party database backends that don't support it can disable it (e.g. SQL Server)
> 
> Given the concern about SQL Server specifically, maybe we should check with the maintainers of [mssql-django](https://github.com/microsoft/mssql-django)? @mShan0 @absci do you have any thoughts here?

Sorry, we're no longer maintaining mssql-django project, not sure what's the future plan.

**Comment by csirmazbendeguz** [comment]
I think it's safe to assume it's not maintained by anyone at the moment then?

@sarahboyce , can we proceed without the feature flag? It's easy to add if someone requests it in the future.

File: tests/foreign_object/test_tuple_lookups.py [hunk 4]
```diff
@@ -126,6 +127,46 @@ def test_in_subquery(self):
             (self.contact_1, self.contact_2, self.contact_5),
         )
 
+    def test_tuple_in_subquery_must_be_query(self):
+        lhs = (F("customer_code"), F("company_code"))
+        # if rhs is any non-Query object with an as_sql() function
```
Comment (sarahboyce) [inline tests/foreign_object/test_tuple_lookups.py:132]:
```suggestion
        # If rhs is any non-Query object with an as_sql() function.
```
(nit)

**Review by sarahboyce — APPROVED** [review]

File: tests/foreign_object/test_tuple_lookups.py (hunk 4 above)
Comment (csirmazbendeguz) [reply tests/foreign_object/test_tuple_lookups.py]: fixed

**Comment by csirmazbendeguz** [comment]
Thank you for the input everybody!

**Pull request merged**